src/lib/drivers/distance_sensor/DistanceSensor.cpp

```diff
@@ -0,0 +1,243 @@
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file DistanceSensor.cpp
+ * Distance Sensor driver base class.
+ */
+
+#include "DistanceSensor.h"
+
+DistanceSensor::DistanceSensor() :
+	ScheduledWorkItem(px4::wq_configurations::hp_default)
+{
+}
+
+DistanceSensor::~DistanceSensor()
+{
+	// Ensure we are truly inactive.
+	stop();
+
+	perf_free(_sample_perf);
+	perf_free(_comms_errors);
+}
+
+int
+DistanceSensor::collect()
+{
+	perf_begin(_sample_perf);
+
+	if (!_collect_phase) {
+		return measure();
+	}
+
+	distance_sensor_s report;
+	report.current_distance = get_distance();
+	report.id               = get_sensor_id();
+	report.max_distance     = _max_distance;
+	report.min_distance     = _min_distance;
+	report.orientation      = get_sensor_orientation();
+	report.signal_quality   = -1;
+	report.timestamp        = hrt_absolute_time();
+	report.type             = distance_sensor_s::MAV_DISTANCE_SENSOR_LASER;
+	report.variance         = get_sensor_variance();
+
+	// Publish the new report.
+	orb_publish(ORB_ID(distance_sensor), _distance_sensor_topic, &report);
+	perf_end(_sample_perf);
+	return PX4_OK;
+}
+
+float
+DistanceSensor::get_field_of_view()
+{
+	return _field_of_view;
+}
+
+uint8_t
+DistanceSensor::get_sensor_id()
+{
+	return 0;
+}
+
+uint8_t
+DistanceSensor::get_sensor_orientation()
+{
+	return _orientation;
+}
+
+uint8_t
+DistanceSensor::get_sensor_variance()
+{
+	return _variance;
+}
+
+int
+DistanceSensor::init()
+{
+	// Intitialize the device.
+	if (dev_init() != OK) {
+		return PX4_ERROR;
+	}
+
+	// Get a publish handle on the range finder topic.
+	distance_sensor_s ds_report = {};
+	_distance_sensor_topic = orb_advertise_multi(ORB_ID(distance_sensor), &ds_report,
+				 &_orb_class_instance, ORB_PRIO_HIGH);
+
+	if (_distance_sensor_topic == nullptr) {
+		PX4_ERR("failed to create distance_sensor object");
+	}
+
+	start();
+	return PX4_OK;
+}
+
+int
+DistanceSensor::measure()
+{
+	return PX4_OK;
+}
+
+int
+DistanceSensor::open_serial_port(const speed_t speed)
```

Comment:
I think handling a serial port is an orthogonal piece.

Reply:
Agreed.  I'll break this out into a separate implementation.